File: app/sbom_file.py
"""Persist a generated SBOM to a timestamped file on disk.

The file is written to the current working directory (documented to be the
repository root) using the pattern ``<timestamp>-snyk-<project_id>-sbom.json``.
That exact file is what the ServiceNow uploader sends (``curl --data-binary``).
"""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

from app.errors import SbomGenerationError
from app.logging import get_logger
from app.snyk_client import SbomDocument


def _timestamp() -> str:
    """Return a filename-safe UTC timestamp: ``YYYYMMDDTHHMMSSZ``."""
    return datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")


def build_filename(project_id: str, *, timestamp: str | None = None) -> str:
    """Build the SBOM filename ``<timestamp>-snyk-<project_id>-sbom.json``."""
    ts = timestamp or _timestamp()
    return f"{ts}-snyk-{project_id}-sbom.json"


def write_sbom(document: SbomDocument, project_id: str, root: Path | None = None) -> Path:
    """Write the SBOM bytes, unmodified, to a timestamped file in ``root``.

    Args:
        document: The generated SBOM to persist.
        project_id: Snyk project id, embedded in the filename.
        root: Target directory (defaults to the current working directory).

    Returns:
        The absolute path of the written file.

    Raises:
        SbomGenerationError: if the file cannot be written.
    """
    logger = get_logger()
    target_dir = (root or Path.cwd()).resolve()
    path = target_dir / build_filename(project_id)

    try:
        path.write_bytes(document.content)
    except OSError as exc:
        raise SbomGenerationError(
            "Failed to write SBOM file to disk.",
            operation="sbom_file.write_sbom",
            url=str(path),
            parsed_message=str(exc),
            next_step="Verify write permissions and available disk space for the target directory.",
        ) from exc

    logger.info("SBOM written to %s (%d bytes)", path, document.size)
    return path

File: go/internal/sbomfile/sbomfile.go
// Package sbomfile persists generated SBOMs to timestamped JSON files.
package sbomfile

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
	"unicode"

	"github.com/snyk-ps/snyk-sbom-to-servicenow/internal/apperr"
	applog "github.com/snyk-ps/snyk-sbom-to-servicenow/internal/logging"
	"github.com/snyk-ps/snyk-sbom-to-servicenow/internal/snyk"
)

// Write writes the document bytes unmodified and returns the absolute path.
func Write(document snyk.Document, projectID, root string, logger *applog.Logger) (string, error) {
	if root == "" {
		var err error
		root, err = os.Getwd()
		if err != nil {
			return "", &apperr.Error{
				Kind:    apperr.KindSBOMGeneration,
				Message: "Unable to determine the SBOM output directory.",
				Op:      "sbomfile.write",
				Detail:  err.Error(),
				Cause:   err,
			}
		}
	}
	absoluteRoot, err := filepath.Abs(root)
	if err != nil {
		return "", &apperr.Error{
			Kind:    apperr.KindSBOMGeneration,
			Message: "Unable to resolve the SBOM output directory.",
			Op:      "sbomfile.write",
			URL:     root,
			Detail:  err.Error(),
			Cause:   err,
		}
	}

	name := fmt.Sprintf(
		"%s-snyk-%s-sbom.json",
		time.Now().UTC().Format("20060102T150405Z"),
		sanitizeProjectID(projectID),
	)
	path := filepath.Join(absoluteRoot, name)
	if err := os.WriteFile(path, document.Content, 0o600); err != nil {
		return "", &apperr.Error{
			Kind:     apperr.KindSBOMGeneration,
			Message:  "Failed to write SBOM file to disk.",
			Op:       "sbomfile.write",
			URL:      path,
			Detail:   err.Error(),
			NextStep: "Verify write permissions and available disk space for the target directory.",
			Cause:    err,
		}
	}
	if logger != nil {
		logger.Info("SBOM written to %s (%d bytes)", path, len(document.Content))
	}
	return path, nil
}

func sanitizeProjectID(projectID string) string {
	var builder strings.Builder
	for _, value := range projectID {
		if unicode.IsLetter(value) || unicode.IsDigit(value) || value == '-' || value == '_' || value == '.' {
			builder.WriteRune(value)
		} else {
			builder.WriteByte('_')
		}
	}
	result := strings.TrimLeft(builder.String(), ".")
	if result == "" {
		return "unknown-project"
	}
	return result
}
